python/ray/air/tests/test_dataset_config.py

```diff
@@ -387,6 +365,127 @@ def checker(shard, results):
     )
     test.fit()
 
+    # TODO(swang): Check that order is randomized on each epoch.
+
+
+@ray.remote
+class DummyTrainer:
+    pass
+
+
+def iterate_twice(trainer, dataset, prep=None, **kwargs):
+    dataset_config = DatasetConfig(split=True, required=True, **kwargs).fill_defaults()
+    spec = DataParallelIngestSpec({"train": dataset_config})
+    datasets = spec.preprocess_datasets(prep, {"train": dataset})
+    it = spec.get_dataset_shards([trainer])[0]["train"]
+
+    results = []
+    for _ in range(2):
+        result = []
+        for batch in it.iter_batches():
+            for row in batch["value"]:
+                result.append(row)
+        results.append(result)
+    return results
+
+
+def test_per_epoch_preprocessor(ray_start_4_cpus):
+    trainer = DummyTrainer.remote()
+    ds = ray.data.range_table(5).fully_executed()
```

Comment:
Without this line, I get this error:
```
ValueError: Dataset is empty or cleared, can't determine the format of the dataset.
```

Reply:
Hmm you get that if you don't have `.fully_executed()`? Do you have the full traceback?

Reply:
Yeah only if I left out `.fully_executed()`. It seems to be an issue with creating the DatasetIterator manually, though, I didn't see the same problem when instantiating the full Trainer. Weirdly it only showed up if I ran once with a stream window size followed by bulk ingest on the same input Dataset. I didn't look into it much since it was incidental but I can look into it more if you think it may be a bug!

Sorry don't have the stacktrace anymore but yeah I think that was where it was raised.

Reply:
Definitely looks like a bug, most likely with the block order randomization hacks, since the schema should definitely be available without triggering full execution.

If you're able to quickly get another backtrace just by deleting that line, that would be super useful! Obviously shouldn't block this effort/PR, but that's a bug we should definitely fix.